### PR TITLE
bugfix/19793-trendline-duplicate-x-values

### LINKED ISSUE #19793: Trendline functionality issue
#### Expected behaviour
Trendline is supposed to be a straight line but it's not. 

#### Actual behaviour
image

#### Live demo with steps to reproduce
https://jsfiddle.net/4cwobzrv/1

#### Product version
11.1.0 (It was ok in the 10.3.3)

#### Affected browser(s)
Any borwser

### PR DESCRIPTION
**Pull request by kamil-musialowski**
Fixed #19793, trendline was not working for duplicate `x` values.

### CODE REVIEW

**Comment by kamil-musialowski** [comment]
@pawelfus I went with the approach for sum of least squares. Everything works just fine imo, however I feel like the code could be a bit cleaner/shorter. If you have any suggestions for further refactoring of that solution, I'm more than happy to implement them.

**Review by pawelfus — CHANGES_REQUESTED** [review]
Perf

**Review by DJTechnoo — APPROVED** [review]
Well done!

File: ts/Stock/Indicators/TrendLine/TrendLineIndicator.ts [hunk 1]
```diff
@@ -107,48 +107,55 @@ class TrendLineIndicator extends SMAIndicator {
...
-            alpha: number,
-            i: number,
-            y: number;
-
-        // Get sums:
-        for (i = 0; i < xValLength; i++) {
-            y = isArray(yVal[i]) ? yVal[i][index] : (yVal[i] as any);
-            sumY += y;
-            sumXY += i * y;
+            index: number = params.index || 3;
```
Comment (pawelfus) [inline ts/Stock/Indicators/TrendLine/TrendLineIndicator.ts:116]: No need for that. If index can be undefined, then it means TS is incorrect (eg DeepPartial is not used)

File: ts/Stock/Indicators/TrendLine/TrendLineIndicator.ts [hunk 2]
```diff
@@ -107,48 +107,55 @@ class TrendLineIndicator extends SMAIndicator {
...
+        let numerator = 0,
+            denominator = 0,
+            xValSum = 0,
+            yValSum = 0,
+            counter = 0;
+
+        // Create an array of consecutive xValues, (don't remove duplicates)
+        for (let i = 0; i < orgXVal.length; i++) {
+            if (i === 0 || orgXVal[i] !== orgXVal[i - 1]) {
+                counter++;
```
Comment (pawelfus) [inline ts/Stock/Indicators/TrendLine/TrendLineIndicator.ts:127]: Just double checking: it means we start the array from x=1, is this expected? Previous solution started from x=0.

**Review by pawelfus — CHANGES_REQUESTED** [review]
Check comments inline. This solution looks promising!

File: ts/Stock/Indicators/TrendLine/TrendLineIndicator.ts (hunk 2 above)
Comment (kamil-musialowski) [reply ts/Stock/Indicators/TrendLine/TrendLineIndicator.ts:127]: Yes, exactly - by default HC xData array starts from 1 (if no value is given), and that what we want to achieve here.

File: ts/Stock/Indicators/TrendLine/TrendLineIndicator.ts (hunk 1 above)
Comment (kamil-musialowski) [reply ts/Stock/Indicators/TrendLine/TrendLineIndicator.ts]: @pawelfus What would you say about the solution in [Added index to TrendLine interface.](https://github.com/highcharts/highcharts/pull/19952/commits/83f26f7fa6ba6cbc1a0c35bf8f10cd0b7b596832) commit?

Comment (pawelfus) [reply ts/Stock/Indicators/TrendLine/TrendLineIndicator.ts]: Looks good!

**Review by bm64 — APPROVED** [review]
LGTM 👍

**Review by pawelfus — APPROVED** [review]
Thanks!

**Review by TorsteinHonsi — APPROVED** [review]

**Pull request merged**